SAO limits: Allow SAOs to exist outside the set 'mapgen limit'. CollisionMoveSimple: Collide with 'ignore' nodes

paramat opened the pull request:
Commit 1:
SAO limits: Allow SAOs to exist outside the set 'mapgen limit'

Commit 2:
CollisionMoveSimple: Collide with 'ignore' nodes 
/////////////

Commit 1:

This is a simpler and more lightweight (personally much preferred) alternative to #6998 

Should be backported to 0.4.17 to help against the recent world-generation server attacks. Server owners need to be able to reduce 'mapgen limit' to limit how much world is generated. This ability is also extremely useful for limiting world database size.

The commit is quite simple as it just removes the per-step SAO position checks.

The intention of 'mapgen limit', which was hmmmm's intention also, is that it only prevents map generation outside a set limit, it should not do anything else. The limit is often changed and reduced for various reasons and the world outside is deliberately visible and accessible to players. It should function normally in all ways.

Previously, this code calculated the exact world edges corresponding to the 'mapgen limit' setting, and deleted SAO entities at these edges. However if a server owner reduces the 'mapgen limit' setting in order to either limit world database size or as a defence against the 'world-generating/copying hacked bot attacks' the world outide the reduced limit then becomes unusable. This was a mistake made in #5889 which was largely my fault as i supported that behaviour.

With this PR, the exact world edges corresponding to the 'mapgen limit' setting are still calculated, but these are now only used to set the maximum new player spawn range which must remain within the set 'mapgen limit'.

The difference from #6998 is that the SAO limits are not calculated and the per-step position checks for SAOs are removed (a slight performance increase), these are now unnecessary because of recent PRs that prevent SAOs entering ignore nodes, and also because of commit 2.

`ServerEnvironment::loadPlayer()` therefore reverts back to using `objectpos_over_limit()`.
//////////////

Commit 2:
'collisionMoveSimple' is used by localplayer, particles with 'collision detection' and CAOs / lua SAOs that have 'physical = true'. Previously there was collision with unloaded areas but not with ignore nodes, causing problems with lua SAOs getting stuck in the ignore at world edge.

VanessaE commented:
I favor outright deleting entities that manage to move beyond whatever the maplimit happens to be.

paramat commented:
Dropped item entities and falling node entites are now deleted on contact with ignore, done in #7024 and #7032 . Those node-resolution solutions are more flexible than deleting at one cubic mapgen limit as they deal with the world edge whereever it is, altering 'mapgen limit' creates a highly irregular world edge which the previous method can't cope with.

Boats now bounce off ignore and carts don't enter it.

VanessaE commented:
If MOBs behave the way carts and boats do, then I can be satisfied with that.

paramat commented:
It's actually the MOB mod's responsibility to make sure MOBs don't enter ignore nodes. Players engine-collide with ignore somehow but MOBs don't because ignore is not 'walkable'. It seems some MOB creators (mobs redo is one) assume MOBs will collide with ignore and make that error.

When mapgen limit is adjusted the world edge becomes highly irregular and the only effective way to stop MOBs falling out of the world is to code them to not enter ignore nodes, just as i have for dropped items, falling nodes and boats.

shivajiva101 commented:
It makes sense to simplify by utilising recent changes so I'm happy to support it especially if it can be backported

paramat commented:
@rubenwardy

celeron55 commented:
It seems kind of cruel to break basically every server that uses this setting without carefully checking their mods for support of the new behavior of this setting. They'd end up having random entities inside the ground, consumping disk space and eating performance.

How do mob mods generally handle movement? If they generally use the engine's built-in collisions, then they could possibly be frozen when being inside an unloaded block or an ingore node or something. No?

paramat commented:
I have found mob mods where the mobs are 'physical' so use engine collision code, but they still enter ignore nodes. So for some reason engine collision code allows entities to enter ignore while players cannot.

Discussing in IRC we suggest making entities collide with ignore just like players. So WIP.

paramat commented:
The check this PR removes was merged just before 0.4.16 stable was released, so yes it could actually affect some servers.

paramat commented:
https://github.com/minetest/minetest/blob/cf0bcebc7620ff2a4bfb44bcbc4909ff9dbe8d24/src/content_sao.cpp#L352
If 'physical = true' SAO uses engine collision code:
```
		if(m_prop.physical){
			aabb3f box = m_prop.collisionbox;
			box.MinEdge *= BS;
			box.MaxEdge *= BS;
			collisionMoveResult moveresult;
			f32 pos_max_d = BS*0.25; // Distance per iteration
			v3f p_pos = m_base_position;
			v3f p_velocity = m_velocity;
			v3f p_acceleration = m_acceleration;
			moveresult = collisionMoveSimple(m_env, m_env->getGameDef(),
					pos_max_d, box, m_prop.stepheight, dtime,
					&p_pos, &p_velocity, p_acceleration,
					this, m_prop.collideWithObjects);

			// Apply results
			m_base_position = p_pos;
			m_velocity = p_velocity;
			m_acceleration = p_acceleration;
```

paramat commented:
The popular mob mod 'mobs redo' sets 'physical = true' for all mobs.

paramat commented:
Relevant code is here https://github.com/minetest/minetest/blob/3face01a202040e4feff3b0936b4aa89c051c98d/src/collision.cpp#L275
There is collision with walkable nodes and with unloaded area but not with ignore.

paramat commented:
2nd commit added that alters 'collisionMoveSimple()' to make ignore collidable, any SAO that is 'physical' should now collide with ignore.

paramat commented:
Testing:

At world edge:
Item entities thrown at the wall of ignore at world's edge hit it and do not enter it.
My 'mesecar' vehicles (physical = true) do not enter ignore, before they would fall off the edge.

At world base:
Dropped item entities and falling node entities do somehow enter ignore and are therefore deleted, due to my recent commit.
My mesecar vehicles rest on ignore and can be driven around on it, the player can exit the vehicle and is not stuck.

paramat commented:
Testing 'mobs redo', mobs no longer fall off the world edge.
At world base they walk on ignore and can jump back up onto world nodes.

![screenshot_20180224_162625](https://user-images.githubusercontent.com/3686677/36632536-a4a6661c-197f-11e8-96a1-97da811a395f.png)

celeron55 commented on src/collision.cpp:
```diff
@@ -280,9 +280,17 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 		MapNode n = map->getNodeNoEx(p, &is_position_valid);
 
 		if (is_position_valid) {
```
Why not just make this ``if (is_position_valid && n.getContent() != CONTENT_IGNORE) {``?

paramat replied:
I considered that, but then 'any position valid' is not set to true for ignore despite ignore being a valid position, this would affect the code below:
```
	// Do not move if world has not loaded yet, since custom node boxes
	// are not available for collision detection.
	if (!any_position_valid) {
		*speed_f = v3f(0, 0, 0);
		return result;
	}
```

paramat replied:
Not sure if that matters or not though, but it might.

celeron55 replied:
I think it's fine to hit it in the case of only ignore being found. As far as movement goes, ignore really is essentially unloaded world. Doing it this way also enables the optimization of skipping the actual collision algorithm in case the object clearly has no business moving anywhere.

Just add `// This also intentionally occurs in the case of the object being positioned solely on CONTENT_IGNORE nodes, no matter where they come from.` before `if (!any_position_valid) {` to make this clear.

paramat replied:
Ok will do.

paramat replied:
Done, will retest.

paramat commented:
Retested dropped item entities, falling node entities, vehicle and mobs, no change in behaviour from previous tests, so this seems ok.
